File: app/javascript/cyvasse/potential_range.js
// Every hex a unit could reach if nothing stood in the way.
//
// A line-for-line port of the legacy walk, amcritchie/Cyvasse
// app/assets/javascripts/hexRange/potentialRange.js. It traces concentric
// rings out from the unit, one "horizontal" step at a time, reading each
// row's width to know which way the rows lean. For every unit but the dragon
// the result is the full hex disc of the given radius; for the dragon it is
// the six straight lines out from its hex. The unit tests hold the port to
// both shapes against an independent cube-coordinate oracle.

import { hexAtXY } from "cyvasse/board";

// Hex indices within `range` of `origin` (a board hex), excluding the origin.
export function potentialRange(origin, range, { dragon = false } = {}) {
  const found = new Set();
  const xPos = origin.x;
  const yPos = origin.y;
  const sizeAt = (x, y) => hexAtXY(x, y)?.size;
  const mark = (x, y) => {
    const hex = hexAtXY(x, y);
    if (hex) found.add(hex.index);
  };

  for (let horizontal = 1; horizontal <= range; horizontal++) {
    let constantUp = 0;
    let constantDown = 0;
    let initialSizeDown = sizeAt(xPos, yPos);
    let initialSizeUp = sizeAt(xPos, yPos);

    mark(xPos - horizontal, yPos);
    mark(xPos + horizontal, yPos);

    for (let vertical = 1; horizontal >= vertical; vertical++) {
      const finalSizeDown = sizeAt(xPos, yPos + vertical);
      const finalSizeUp = sizeAt(xPos, yPos - vertical);

      // An undefined size (off the board) compares false, as it did in jQuery.
      if (initialSizeDown < finalSizeDown) constantDown += 1;
      if (initialSizeUp < finalSizeUp) constantUp += 1;

      for (const [constant, up] of [[constantDown, vertical], [constantUp, -vertical]]) {
        if (dragon) {
          if (horizontal <= vertical) {
            const xx = xPos - (horizontal - vertical - constant);
            mark(xx, yPos + up);
            mark(xx - horizontal, yPos + up);
          }
        } else if (horizontal <= vertical) {
          for (let row = 0; row <= horizontal; row++) {
            mark(xPos - (horizontal + row - vertical - constant), yPos + up);
          }
        } else {
          mark(xPos - (horizontal - constant), yPos + up);
          mark(xPos + (horizontal + constant - vertical), yPos + up);
        }
      }

      initialSizeDown = finalSizeDown;
      initialSizeUp = finalSizeUp;
    }
  }
  return found;
}

File: app/models/cyvasse_rules/potential_range.rb
module CyvasseRules
  # Every hex a unit could reach if nothing stood in the way: a line-for-line
  # mirror of app/javascript/cyvasse/potential_range.js (itself a port of the
  # legacy hexRange/potentialRange.js). For every unit but the dragon it is
  # the full hex disc of the radius; for the dragon, six straight lines.
  #
  # JavaScript compares `undefined < n` as false where Ruby raises, so every
  # size comparison goes through `lt`.
  module PotentialRange
    def self.call(origin, range, dragon: false)
      found = Set.new
      x_pos = origin.x
      y_pos = origin.y
      size_at = ->(x, y) { Board.hex_at_xy(x, y)&.size }
      mark = ->(x, y) { (hex = Board.hex_at_xy(x, y)) && found << hex.index }
      lt = ->(a, b) { !a.nil? && !b.nil? && a < b }

      (1..range).each do |horizontal|
        constant_up = 0
        constant_down = 0
        initial_size_down = size_at.(x_pos, y_pos)
        initial_size_up = size_at.(x_pos, y_pos)

        mark.(x_pos - horizontal, y_pos)
        mark.(x_pos + horizontal, y_pos)

        (1..horizontal).each do |vertical|
          final_size_down = size_at.(x_pos, y_pos + vertical)
          final_size_up = size_at.(x_pos, y_pos - vertical)

          constant_down += 1 if lt.(initial_size_down, final_size_down)
          constant_up += 1 if lt.(initial_size_up, final_size_up)

          [ [ constant_down, vertical ], [ constant_up, -vertical ] ].each do |constant, up|
            if dragon
              if horizontal <= vertical
                xx = x_pos - (horizontal - vertical - constant)
                mark.(xx, y_pos + up)
                mark.(xx - horizontal, y_pos + up)
              end
            elsif horizontal <= vertical
              (0..horizontal).each { |row| mark.(x_pos - (horizontal + row - vertical - constant), y_pos + up) }
            else
              mark.(x_pos - (horizontal - constant), y_pos + up)
              mark.(x_pos + (horizontal + constant - vertical), y_pos + up)
            end
          end

          initial_size_down = final_size_down
          initial_size_up = final_size_up
        end
      end
      found
    end
  end
end
